### PR TITLE
[mono] Avoid an assertion when cattrs with array parameters are const…

### LINKED ISSUE #57361: [mono] FullAOT runtime test failures: condition `out_obj' not met
These tests crash--for the same reason, with nearly identical backtraces--on FullAOT compilation attempts:
- `Interop/SuppressGCTransition/SuppressGCTransitionTest/SuppressGCTransitionTest.dll` 
- `JIT/Directed/callconv/CdeclMemberFunction/CdeclMemberFunctionTest/CdeclMemberFunctionTest.dll`
- `JIT/Directed/callconv/PlatformDefaultMemberFunction/PlatformDefaultMemberFunctionTest/PlatformDefaultMemberFunctionTest.dll`
- `JIT/Directed/callconv/StdCallMemberFunction/StdCallMemberFunctionTest/StdCallMemberFunctionTest.dll`
- `JIT/Directed/callconv/ThisCall/ThisCallTest/ThisCallTest.dll`

The compilation flags used are:
```
MONO_ENV_OPTIONS="--aot=full,llvm,llvm-path=$MONO_PFX,mattr=sse4.2,mattr=popcnt,mattr=lzcnt,mattr=bmi,mattr=bmi2,mattr=pclmul,mattr=aes"
```

The following output is produced:

```
* Assertion at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:595, condition `out_obj' not met


=================================================================
        Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
        Native stacktrace:
=================================================================
        0x7f8bf42b32da - Unknown
        0x7f8bf42b3251 - Unknown
        0x7f8bf41c0178 - Unknown
        0x7f8bf42b29c2 - Unknown
        0x7f8bf6177730 - Unknown
        0x7f8bf5cba7bb - Unknown
        0x7f8bf5ca5535 - Unknown
        0x7f8bf43723e2 - Unknown
        0x7f8bf3fd37ae - Unknown
        0x7f8bf3fd04dc - Unknown
        0x7f8bf437305a - Unknown
        0x7f8bf4372a8e - Unknown
        0x7f8bf4372db4 - Unknown
        0x7f8bf4372e2c - Unknown
        0x7f8bf3f80511 - Unknown
        0x7f8bf3f7ef39 - Unknown
        0x7f8bf418986e - Unknown
        0x7f8bf41806a9 - Unknown
        0x7f8bf417d43e - Unknown
        0x7f8bf4173d33 - Unknown
        0x7f8bf417271f - Unknown
        0x7f8bf422e7d7 - Unknown
        0x7f8bf40a2512 - Unknown
        0x55ff93b824ab - Unknown
        0x7f8bf5ca709b - Unknown
        0x55ff93b7e41a - Unknown

=================================================================
        Telemetry Dumper:
=================================================================
No threads attached to runtime.

=================================================================
        External Debugger Dump:
=================================================================
[New LWP 60814]
BFD: /usr/lib/debug/.build-id/d3/583c742dd47aaa860c5ae0c0c5bdbcd2d54f61.debug: unable to initialize decompress status for section .debug_aranges
BFD: /usr/lib/debug/.build-id/d3/583c742dd47aaa860c5ae0c0c5bdbcd2d54f61.debug: unable to initialize decompress status for section .debug_aranges
warning: File "/usr/lib/debug/.build-id/d3/583c742dd47aaa860c5ae0c0c5bdbcd2d54f61.debug" has no build-id, file skipped
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
BFD: /usr/lib/debug/.build-id/88/5dda4b4a5cea600e7b5b98c1ad86996c8d2299.debug: unable to initialize decompress status for section .debug_aranges
BFD: /usr/lib/debug/.build-id/88/5dda4b4a5cea600e7b5b98c1ad86996c8d2299.debug: unable to initialize decompress status for section .debug_aranges
warning: File "/usr/lib/debug/.build-id/88/5dda4b4a5cea600e7b5b98c1ad86996c8d2299.debug" has no build-id, file skipped
BFD: /usr/lib/debug/.build-id/18/b9a9a8c523e5cfe5b5d946d605d09242f09798.debug: unable to initialize decompress status for section .debug_aranges
BFD: /usr/lib/debug/.build-id/18/b9a9a8c523e5cfe5b5d946d605d09242f09798.debug: unable to initialize decompress status for section .debug_aranges
warning: File "/usr/lib/debug/.build-id/18/b9a9a8c523e5cfe5b5d946d605d09242f09798.debug" has no build-id, file skipped
BFD: /usr/lib/debug/.build-id/f2/5dfd7b95be4ba386fd71080accae8c0732b711.debug: unable to initialize decompress status for section .debug_aranges
BFD: /usr/lib/debug/.build-id/f2/5dfd7b95be4ba386fd71080accae8c0732b711.debug: unable to initialize decompress status for section .debug_aranges
warning: File "/usr/lib/debug/.build-id/f2/5dfd7b95be4ba386fd71080accae8c0732b711.debug" has no build-id, file skipped
BFD: /usr/lib/debug/.build-id/5d/cf98ad684962be494af28a1051793fd39e4ebc.debug: unable to initialize decompress status for section .debug_aranges
BFD: /usr/lib/debug/.build-id/5d/cf98ad684962be494af28a1051793fd39e4ebc.debug: unable to initialize decompress status for section .debug_aranges
warning: File "/usr/lib/debug/.build-id/5d/cf98ad684962be494af28a1051793fd39e4ebc.debug" has no build-id, file skipped
0x00007f8bf61770ca in __waitpid (pid=60816, stat_loc=0x7ffcc7fa7b7c, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
30      ../sysdeps/unix/sysv/linux/waitpid.c: No such file or directory.
  Id   Target Id         Frame
* 1    Thread 0x7f8bf5c7e740 (LWP 60813) "corerun" 0x00007f8bf61770ca in __waitpid (pid=60816, stat_loc=0x7ffcc7fa7b7c, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
  2    Thread 0x7f8bf1fff700 (LWP 60814) "SGen worker" futex_wait_cancelable (private=0, expected=0, futex_word=0x7f8bf5c43d38 <work_cond+40>) at ../sysdeps/unix/sysv/linux/futex-internal.h:88

Thread 2 (Thread 0x7f8bf1fff700 (LWP 60814)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0x7f8bf5c43d38 <work_cond+40>) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f8bf5c43ce8 <lock>, cond=0x7f8bf5c43d10 <work_cond>) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7f8bf5c43d10 <work_cond>, mutex=0x7f8bf5c43ce8 <lock>) at pthread_cond_wait.c:655
#3  0x00007f8bf408ea4d in mono_os_cond_wait (cond=0x7f8bf5c43d10 <work_cond>, mutex=0x7f8bf5c43ce8 <lock>) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/../../mono/utils/mono-os-mutex.h:219
#4  0x00007f8bf408f5bf in get_work (worker_index=0, work_context=0x7f8bf1ffeee0, do_idle=0x7f8bf1ffeed4, job=0x7f8bf1ffeec8) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/sgen/sgen-thread-pool.c:167
#5  0x00007f8bf408e5ab in thread_func (data=0x0) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/sgen/sgen-thread-pool.c:198
#6  0x00007f8bf616cfa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#7  0x00007f8bf5d7c4cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 1 (Thread 0x7f8bf5c7e740 (LWP 60813)):
#0  0x00007f8bf61770ca in __waitpid (pid=60816, stat_loc=0x7ffcc7fa7b7c, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
#1  0x00007f8bf42b3773 in dump_native_stacktrace (signal=0x7f8bf34b9247 "SIGABRT", mctx=0x7ffcc7fa8788) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/mini-posix.c:981
#2  0x00007f8bf42b3251 in mono_dump_native_crash_info (signal=0x7f8bf34b9247 "SIGABRT", mctx=0x7ffcc7fa8788, info=0x7ffcc7fa8a70) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/mini-posix.c:1025
#3  0x00007f8bf41c0178 in mono_handle_native_crash (signal=0x7f8bf34b9247 "SIGABRT", mctx=0x7ffcc7fa8788, info=0x7ffcc7fa8a70) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/mini-exceptions.c:3382
#4  0x00007f8bf42b29c2 in sigabrt_signal_handler (_dummy=6, _info=0x7ffcc7fa8a70, context=0x7ffcc7fa8940) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/mini-posix.c:227
#5  <signal handler called>
#6  0x00007f8bf5cba7bb in raise () from /lib/x86_64-linux-gnu/libc.so.6
#7  0x00007f8bf5ca5535 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#8  0x00007f8bf43723e2 in monoeg_assert_abort () at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/eglib/goutput.c:57
#9  0x00007f8bf3fd37ae in mono_log_write_logfile (log_domain=0x0, level=G_LOG_LEVEL_ERROR, hdr=0, message=0x55ff94ac0360 "* Assertion at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:595, condition `out_obj' not met\n") at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/utils/mono-log-common.c:136
#10 0x00007f8bf3fd04dc in structured_log_adapter (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, message=0x55ff94ac0360 "* Assertion at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:595, condition `out_obj' not met\n", user_data=0x0) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/utils/mono-logger.c:461
#11 0x00007f8bf437305a in monoeg_g_logstr (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, msg=0x55ff94ac0360 "* Assertion at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:595, condition `out_obj' not met\n") at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/eglib/goutput.c:151
#12 0x00007f8bf4372a8e in monoeg_g_logv_nofree (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, format=0x7f8bf35300d5 "* Assertion at %s:%d, condition `%s' not met\n", args=0x7ffcc7fa94b0) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/eglib/goutput.c:166
#13 0x00007f8bf4372db4 in monoeg_assertion_message (format=0x7f8bf35300d5 "* Assertion at %s:%d, condition `%s' not met\n") at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/eglib/goutput.c:207
#14 0x00007f8bf4372e2c in mono_assertion_message (file=0x7f8bf3232b94 "/home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c", line=595, condition=0x7f8bf31f9c2b "out_obj") at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/eglib/goutput.c:226
#15 0x00007f8bf3f80511 in load_cattr_value (image=0x55ff94a4cb30, t=0x55ff94a9e898, out_obj=0x0, p=0x7f8bf61a3076 "", boundp=0x7f8bf61a3076 "", end=0x7ffcc7fa9810, error=0x7ffcc7fa9c00) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:595
#16 0x00007f8bf3f7ef39 in mono_reflection_create_custom_attr_data_args_noalloc (image=0x55ff94a4cb30, method=0x55ff94a9e8b8, data=0x7f8bf61a2fdf "\001", len=151, typed_args_out=0x7ffcc7fa9b38, named_args_out=0x7ffcc7fa9b30, num_named_args=0x7ffcc7fa9b24, named_arg_info=0x7ffcc7fa9b28, error=0x7ffcc7fa9c00) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/metadata/custom-attrs.c:1311
#17 0x00007f8bf418986e in add_wrappers (acfg=0x55ff94aa6320) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/aot-compiler.c:5127
#18 0x00007f8bf41806a9 in collect_methods (acfg=0x55ff94aa6320) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/aot-compiler.c:12292
#19 0x00007f8bf417d43e in mono_compile_assembly (ass=0x55ff94aa3f60, opts=374434303, aot_options=0x55ff94a20470 "full,llvm,llvm-path=/home/imran/ms/dnr-11.x/runtime/artifacts/bin/mono/Linux.x64.Release,mattr=sse4.2,mattr=popcnt,mattr=lzcnt,mattr=bmi,mattr=bmi2,mattr=pclmul,mattr=aes", global_aot_state=0x7ffcc7faa2c8) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/aot-compiler.c:14053
#20 0x00007f8bf4173d33 in main_thread_handler (user_data=0x7ffcc7faa620) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/driver.c:1431
#21 0x00007f8bf417271f in mono_main (argc=3, argv=0x55ff94a3dd80) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/driver.c:2679
#22 0x00007f8bf422e7d7 in monovm_execute_assembly (argc=0, argv=0x55ff94a17140, managedAssemblyPath=0x55ff94a17200 "/home/imran/ms/dnr-11.x/runtime/artifacts/tests/coreclr/Linux.x64.Release/Interop/SuppressGCTransition/SuppressGCTransitionTest/SuppressGCTransitionTest.dll", exitCode=0x7ffcc7faa7ac) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/monovm.c:270
#23 0x00007f8bf40a2512 in coreclr_execute_assembly (hostHandle=0x0, domainId=0, argc=0, argv=0x55ff94a17140, managedAssemblyPath=0x55ff94a17200 "/home/imran/ms/dnr-11.x/runtime/artifacts/tests/coreclr/Linux.x64.Release/Interop/SuppressGCTransition/SuppressGCTransitionTest/SuppressGCTransitionTest.dll", exitCode=0x7ffcc7faa7ac) at /home/imran/ms/dnr-11.x/runtime/src/mono/mono/mini/main-core.c:81
#24 0x000055ff93b824ab in run (config=...) at /home/imran/ms/dnr-11.x/runtime/src/coreclr/hosts/corerun/corerun.cpp:372
#25 main (argc=<optimized out>, argv=<optimized out>) at /home/imran/ms/dnr-11.x/runtime/src/coreclr/hosts/corerun/corerun.cpp:554
```

### PR DESCRIPTION
**Pull request by vargaz**
…ructed during AOT.

Fixes https://github.com/dotnet/runtime/issues/57361.

### CODE REVIEW

File: src/mono/mono/metadata/custom-attrs.c [hunk 1]
```diff
@@ -574,34 +580,44 @@ MONO_RESTORE_WARNING
...
+					mono_array_setref_internal (arr, i, item);
+				}
+				HANDLE_FUNCTION_RETURN ();
+			} else {
+				for (i = 0; i < alen; i++) {
+					MonoObject *item = NULL;
+					load_cattr_value (image, m_class_get_byval_arg (tklass), &item, p, boundp, &p, error);
+					if (!is_ok (error))
+						return NULL;
+				}
```
Comment (lambdageek) [inline src/mono/mono/metadata/custom-attrs.c:610]: This doesn't really fix the root issue (https://github.com/dotnet/runtime/issues/52977), it just reads the values and throws them away.

I think @SamMonoRT wanted @ivanpovazan to work on a proper fix.

We do care about this on non-windows platforms, by the way.  we want to support `[UnmanagedCallersOnly (CallConvs = new Type[] {CallConvCdecl, CallConvSuppressGCTransition})]`, for example.

I guess this is ok if it unblocks https://github.com/dotnet/runtime/issues/57361 if we want to do that quickly for now

Comment (SamMonoRT) [reply src/mono/mono/metadata/custom-attrs.c:610]: This should be ok for a quick fix to unblock 57361. @ivanpovazan will work on a proper fix and maybe change this temporary fix during that time. He'll start investigating this sometime this week.

**Review by lambdageek — APPROVED** [review]

**Comment by vargaz** [comment]
Failures are unrelated.

**Comment by ManickaP** [comment]
Does this mean that I can get rid of all the Mono AOT exclusions in https://github.com/dotnet/runtime/pull/69009? If yes, then thanks a lot!!!!

**Comment by vargaz** [comment]
Its possible if they failed with the same error.

**Pull request merged**